### PR TITLE
support extension ckan-galleries

### PR DESCRIPTION
**Pull request by florianm**
Addressing #209 

Three hurdles taken, one remaining. I'm posting this PR as point for discussion, although it's not complete.

This should fix the missing mysql_config reported in #201 by adding libmysqlclient-dev to the apt-get install steps in the Dockerfile.

Two additional version conflicts (requests and six, ckan vs ckan-galleries) have to be handled manually until ckan upgrades their versions.

One mean Docker permissions bug remains unsolved.

Steps taken:
```
# create /var/www/projects/datacats so python package requests won't break pip
mkvirtualenv datacats

# Override datacats data dir location to my backed up, bind-mounted btrfs partition at /var/www:
# (before creating a datacats environment)
ln -sf /var/www/projects/datacats/data ~/.datacats

# pull my datacats fork
/var/www/projects/datacats$ git clone git@github.com:florianm/datacats.git
/var/www/projects/datacats/datacats$ python setup.py install

# create a datacats environment
/var/www/projects/datacats$ datacats create modena
/var/www/projects/datacats/modena/

# clone and install datacats into the new environment
/var/www/projects/datacats/modena$ git clone https://github.com/DataShades/ckan-galleries.git
python setup.py install

# modify ckan's requirements to use a ckan-galleries-compatible requests==2.7.0 and six==1.8.0
/var/www/projects/datacats/modena$ vim ckan/requirements.txt

# add ckan-galleries extension name "dfmp" to development.ini
vim /var/www/projects/datacats/modena/development.ini 

# datacats install will not fail anymore on missing mysql_config (fixed through this PR), or
# version conflicts for requests and six (fixed in steps above, to be included in this PR)
datacats install modena

# still running into Docker permissions bug: can't write to folders inside container
datacats start modena --address=0.0.0.0
```
Docker permissions bug: 
ckan-galleries creates a flickrapi instance, which creates a local (=inside the docker container) folder to store oauth tokens. The [docker-bug](https://github.com/docker/docker/issues/1295) is prorbably to do with stricter aufs permissions restricting their way up through the snapshot layers and strangling  flickrapi's attempt to create a folder `/root/.flickrapi`. This one's gonna be messy.

Re package dependencies, where does a datacats envorinment's `ckan/requirements.txt` come from, and can I update the requirements there?

### CODE REVIEW

**Comment by JackMc** [comment]
requirements.txt comes from CKAN itself when we build the web image: https://github.com/ckan/ckan/blob/release-v2.3-latest/requirements.txt

**Comment by JackMc** [comment]
This looks good, though!!

**Comment by wardi** [comment]
@florianm ckan is part of your source directory so you're free to modify any part of it, including its requirements.txt file. Then run `datacats install` and you should have the deps you need

**Comment by florianm** [comment]
Coolio. I just asked to see whether I can include the two updated requirements to the pull request. It seems however that the docker permission problem might require swapping aufs for a different file system. Will have to sleep over that one, also as it's gmt+8 over here.

**Comment by wardi** [comment]
It's strange that anything is being created under /root/. Could that code be running in the wrong place? When running commands with e.g. `datacats shell` the home folder is your environment source folder which is a volume mount to your host filesystem.

**Comment by florianm** [comment]
@wardi ckan-galleries uses flickrapi, which at runtime creates a folder ~/.flickrapi to store auth tokens.

I can install ckan-galleries fine, but enabling its extension `dfmp` in the development.ini and starting the containers will throw said permission error:

```(datacats)florianm@aws-eco-002:/var/www/projects/datacats$ datacats start modena --address=0.0.0.0
Starting web server at http://0.0.0.0:5687/ ...
----------------------------------------
datacats: problem was encountered:
Error while starting web container:
Traceback (most recent call last):
  File "/usr/lib/ckan/bin/paster", line 9, in <module>
    load_entry_point('PasteScript==1.7.5', 'console_scripts', 'paster')()
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/command.py", line 104, in run
    invoke(command, command_name, options, args[1:])
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/command.py", line 143, in invoke
    exit_code = runner.run(args)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/command.py", line 238, in run
    result = self.command()
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/serve.py", line 284, in command
    relative_to=base, global_conf=vars)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/script/serve.py", line 321, in loadapp
    **kw)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 247, in loadapp
    return loadobj(APP, uri, name=name, **kw)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 272, in loadobj
    return context.create()
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 710, in create
    return self.object_type.invoke(self)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/deploy/loadwsgi.py", line 146, in invoke
    return fix_call(context.object, context.global_conf, **context.local_conf)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/paste/deploy/util.py", line 56, in fix_call
    val = callable(*args, **kw)
  File "/project/ckan/ckan/config/middleware.py", line 58, in make_app
    load_environment(conf, app_conf)
  File "/project/ckan/ckan/config/environment.py", line 232, in load_environment
    p.load_all(config)
  File "/project/ckan/ckan/plugins/core.py", line 134, in load_all
    load(*plugins)
  File "/project/ckan/ckan/plugins/core.py", line 148, in load
    service = _get_service(plugin)
  File "/project/ckan/ckan/plugins/core.py", line 254, in _get_service
    return plugin.load()(name=plugin_name)
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/pkg_resources.py", line 2048, in load
    entry = __import__(self.module_name, globals(),globals(), ['__name__'])
  File "/project/ckan-galleries/ckanext/dfmp/plugin.py", line 10, in <module>
    from ckanext.dfmp.actions.background import *
  File "/project/ckan-galleries/ckanext/dfmp/actions/background.py", line 19, in <module>
    import ckanext.dfmp.scripts.flickr_import as flickr_import
  File "/project/ckan-galleries/ckanext/dfmp/scripts/flickr_import.py", line 13, in <module>
    flickr = flickrapi.FlickrAPI(api_key, api_secret, format='parsed-json')
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/flickrapi/core.py", line 199, in __init__
    self.token_cache = tokencache.OAuthTokenCache(api_key, username or '')
  File "/usr/lib/ckan/local/lib/python2.7/site-packages/flickrapi/tokencache.py", line 132, in __init__
    os.makedirs(self.path)
  File "/usr/lib/ckan/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/root/.flickr'
Starting subprocess with file monitor
```

**Comment by deniszgonjanin** [comment]
@florianm the web containers are treated as read-only, so that they can be thrown away at any time with no consequence.

The fix is probably to store the auth tokens in a docker volume, preferably the same volume that files and ckan sessions are already stored in.

**Comment by florianm** [comment]
@deniszgonjanin would you happen to know the path to that volume? FlickrAPI will accept a custom path  for its tokencache folder. I could fork ckan-galleries and parameterise the custom path into a config parameter.

**Comment by wardi** [comment]
We must not be running paster shell as the right user with the right settings. I'll investigate. currently the only places you can write when running the server is under /var/www/storage

**Comment by deniszgonjanin** [comment]
What about the place CKAN stores repoze sessions - aren't those also stored on the filesystem?

**Comment by florianm** [comment]
I'll see whether I can get the FlickAPI to write to /var/www/storage in the mean time.

**Comment by wardi** [comment]
@florianm You can add your own directories to https://github.com/datacats/datacats/blob/master/datacats/environment.py#L809-L811  e.g.  `'/some/local/path': '/root'` would make root writable inside the container and store the files in /some/local/path on your host.

@deniszgonjanin good question! I don't see a bind mount for those.

**Comment by deniszgonjanin** [comment]
I think MC @JackMc looked into those sessions a while ago, he might know where they are stored

**Comment by florianm** [comment]
No luck yet. Added path to environment.py:809ff, ran `python.py setup install` in datacats source dir, ran `datacats install modena` on my environment with ckan-galleries, ran `datacats start modena --address=0.0.0.0` getting the same stack trace as above. Either the mapping hasn't worked, or the changes aren't propagated, or the too strict permission preventing me from writing to /root/ even propagates to mapped out volumes.

**Comment by florianm** [comment]
Here's the datacats environment's data folder:
```
drwx------ 1 florianm domain users  94 Jun 18 03:29 ./
drwx------ 1 florianm domain users  14 Jun 16 06:17 ../
drwxr-xr-x 1 florianm domain users   0 Jun 16 06:17 data/
drwxr-xr-x 1 ubuntu   www-data      14 Jun 16 06:18 files/
-rw-r--r-- 1 florianm domain users 206 Jun 16 06:17 passwords.ini
drwx------ 1      999 root         386 Jun 18 02:40 postgres/
drwxr-xr-x 1 root     root          14 Jun 18 03:33 root/
drwxr-xr-x 1 florianm domain users  60 Jun 16 07:28 run/
drwxr-xr-x 1 florianm domain users   0 Jun 16 06:17 search/
drwxr-xr-x 1 root     root           8 Jun 16 06:17 solr/
```
So, creating the `/root/.flickr` folder worked as it turns up here. Although, the permissions are different from e.g. `files/`.

**Comment by wardi** [comment]
@florianm try setting its permissions the same as files with `chown --reference=files root` or similar

**Comment by florianm** [comment]
Hang on, had the folders the wrong way around. Too much Thursday, not enough coffee.

**Comment by florianm** [comment]
So, ckan-galleries does work (the CKAN starts up, pending GUI-testing of all functions).
I'll try and summarise the steps (-> section in docs as soon as ckan and cckan-galleries merge PRs).

```
# create a datacats environment
/var/www/projects/datacats$ datacats create modena
/var/www/projects/datacats/modena/

# clone and install ckan-galleries into the new environment
# NOTE this will change to datashades/ckan-galleries as soon as my PR is merged
/var/www/projects/datacats/modena$ git clone https://github.com/florianm/ckan-galleries.git

# NOTE this won't be necessary if ckan update their requirement versions for requests and six
# modify ckan's requirements to use a ckan-galleries-compatible requests==2.7.0 and six==1.8.0
/var/www/projects/datacats/modena$ vim ckan/requirements.txt

# add ckan-galleries extension name "dfmp" to development.ini
vim /var/www/projects/datacats/modena/development.ini 

# install environment
datacats install modena

# run local server for public access
datacats start modena --address=0.0.0.0
```

**Comment by wardi** [comment]
If you want to get a little fancier you can replace the datacats create, git clone, config editing and datacats install steps with a "datacats init" in a source directory that already has those files set up. We've been experimenting with git submodules to distribute source directories, but rsync or similar work too.

You should also run "datacats start" with -p if you're using this in production. That runs the site under apache with debug=false.

I believe all we need to do on the datacats side to fix this issue is make sure the HOME env var is set to /var/www/storage within the container running the web server. this should be an easy fix.

File: datacats/environment.py [hunk 1]
```diff
@@ -571,6 +571,17 @@ def fix_storage_permissions(self):
...
+        """
+        Set the owner of the flickrapi-tokencache folder to www-data container user
+
+        This folder is required for extensions using the flickrapi,
+        e.g. ckan-galleries
+        """
+        web_command(
+            command='/bin/chown -R www-data: /root/.flickr',
+            rw={self.sitedir + '/flickrapi-tokencache': '/root/.flickr'})
+
```
Comment (wardi) [inline datacats/environment.py]: if we set the $HOME setting properly when running the server this and the extra bind mounts won't be necessary

**Comment by JackMc** [comment]
@florianm Please check if #225 addresses the problem here

**Comment by florianm** [comment]
Checking using latest datacats master, speed bump at postgis install:
```
(datacats)florianm@aws-eco-002:/var/www/projects/datacats$ datacats create ali
Creating environment "ali/primary"...........
Installing ckan
Installing ckanext-alitheme
Installing datapusher
Installing ckan requirements
Installing datapusher requirements
Initializing database
----------------------------------------
datacats: problem was encountered:
Something that should not have happened happened when attempting to run this command:
     datacats create ali

It is seems to be a bug.
Please report this issue to us by creating an issue ticket at

    https://github.com/datacats/datacats/issues

so that we would be able to look into that and fix the issue.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Please quote this traceback when reporting this error:
Traceback (most recent call last):
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/datacats-1.1.0dev1-py2.7.egg/datacats/cli/main.py", line 93, in main
    return command_fn(opts)
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/datacats-1.1.0dev1-py2.7.egg/datacats/cli/create.py", line 53, in create
    datapusher=not opts['--no-datapusher'],
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/datacats-1.1.0dev1-py2.7.egg/datacats/cli/create.py", line 97, in create_environment
    log_syslog=log_syslog)
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/datacats-1.1.0dev1-py2.7.egg/datacats/cli/create.py", line 215, in finish_init
    environment.install_postgis_sql()
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/datacats-1.1.0dev1-py2.7.egg/datacats/environment.py", line 351, in install_postgis_sql
    links={self._get_container_name('postgres'): 'db'},
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/datacats-1.1.0dev1-py2.7.egg/datacats/docker.py", line 172, in web_command
    volumes_from=volumes_from)
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/docker/utils/decorators.py", line 15, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/docker/client.py", line 1063, in start
    self._raise_for_status(res)
  File "/var/www/virtualenvs/datacats/local/lib/python2.7/site-packages/docker/client.py", line 122, in _raise_for_status
    raise errors.APIError(e, response, explanation=explanation)
APIError: 404 Client Error: Not Found ("Cannot start container 14682c2c8206dc71c851ef9b3c33c9f2d98c31ce6a55e1d9b4c2d75cc3ee5c56: [8] System error: exec: "\"/scripts/install_postgis.sh\"": stat "/scripts/install_postgis.sh": no such file or directory")

```

Will investigate more!

**Comment by JackMc** [comment]
Thanks for the report, do you have any more details regarding what conditions cause this to happen?

**Comment by florianm** [comment]
Tried again with latest datacats master and latest images, although on my local machine (not the server I reported the error two comments above). Got a new warning about the environment name requiring at least 5 characters (mine was 3 - I'm running out of longer coffee blend names to name my test environments). As I did not encounter the Client Error I reported above I'd say that was due to an outdated datacats install. I still need to test for the ckan-galleries/flickrapi permission bug.

**Comment by JackMc** [comment]
I think this is already fixed in Datacats, if not please comment or reopen

**Pull request closed without merging**